app/src/main/java/org/wikipedia/notifications/PollNotificationService.kt

```diff
@@ -0,0 +1,106 @@
+        return true
+    }
+
+    override fun onStopJob(p0: JobParameters?): Boolean {
+        jobParameters = p0
+        return true
+    }
+
+    private fun pollNotifications(context: Context) {
+        CoroutineScope(Dispatchers.Default).launch(CoroutineExceptionHandler { _, t ->
+            if (t is MwException && t.error.title == "login-required") {
+                assertLoggedIn()
+            }
+            L.e(t)
+            pollNotificationJobFinished()
+        }) {
+            val response = ServiceFactory.get(WikipediaApp.getInstance().wikiSite).lastUnreadNotification()
+            val lastNotificationTime =
+                response.query?.notifications?.list?.maxOfOrNull { it.utcIso8601 }.orEmpty()
+            if (lastNotificationTime > Prefs.remoteNotificationsSeenTime) {
+                Prefs.remoteNotificationsSeenTime = lastNotificationTime
+                retrieveNotifications(context)
+            }
+        }
+    }
+
+    private suspend fun retrieveNotifications(context: Context) {
+        NotificationPollBroadcastReceiver.DBNAME_WIKI_SITE_MAP.clear()
+        NotificationPollBroadcastReceiver.DBNAME_WIKI_NAME_MAP.clear()
+        val response =
+            ServiceFactory.get(WikipediaApp.getInstance().wikiSite).unreadNotificationWikis()
+        val wikiMap = response.query!!.unreadNotificationWikis
+        val wikis = mutableListOf<String>()
+        wikis.addAll(wikiMap!!.keys)
+        for (dbName in wikiMap.keys) {
+            if (wikiMap[dbName]!!.source != null) {
+                NotificationPollBroadcastReceiver.DBNAME_WIKI_SITE_MAP[dbName] =
+                    WikiSite(wikiMap[dbName]!!.source!!.base)
+                NotificationPollBroadcastReceiver.DBNAME_WIKI_NAME_MAP[dbName] =
+                    wikiMap[dbName]!!.source!!.title
+            }
+        }
+        getFullNotifications(context, wikis)
+    }
+
+    private suspend fun getFullNotifications(context: Context, foreignWikis: List<String?>) {
+        val notificationRepository = NotificationRepository(AppDatabase.getAppDatabase().notificationDao())
+
+        val response = ServiceFactory.get(WikipediaApp.getInstance().wikiSite)
+            .getAllNotifications(if (foreignWikis.isEmpty()) "*" else foreignWikis.joinToString("|"), "!read", null)
+        response.query?.notifications?.list?.let {
+            notificationRepository.insertNotifications(it)
+            NotificationPollBroadcastReceiver.onNotificationsComplete(context, it)
+        }
+    }
+
+    private fun assertLoggedIn() {
+        // Attempt to get a dummy CSRF token, which should automatically re-log us in explicitly,
+        // and should automatically log us out if the credentials are no longer valid.
+        CsrfTokenClient(WikipediaApp.getInstance().wikiSite).token
+            .subscribeOn(Schedulers.io())
+            .subscribe()
+    }
+
+    companion object {
+        var pollNotificationService = PollNotificationService()
```

Comment:
Instances of `Service` classes are created by the system. Keeping our own instance of a Service class is not correct.